gene/gene.go

```diff
@@ -0,0 +1,178 @@
+	cdnaSeq      []dna.Base   // The cDNA sequence of the gene from 5' to 3'.
+	featureArray []Feature    // FeatureArray is a slice with len(featureArray) = len(genomeSeq). The index of featureArray corresponds to the same index of genomeSeq. featureArray denotes the features listed above as negative values, or the cDNA pos using all values > 0.
+	orig         goGeneBackup // Copy of initial Gene state to enable the Reset() function.
+	changeLog    []diff       // Log of any mutations that have been performed on the Gene to enable the Reset() function.
+}
+
+type goGeneBackup struct {
+	startPos     int
+	cdsStarts    []int
+	cdsEnds      []int
+	genomeSeq    []dna.Base
+	cdnaSeq      []dna.Base
+	featureArray []Feature
+}
+
+type diff struct {
+	genomePos int
+	removed   []dna.Base
+	added     []dna.Base
+}
+
+type EffectPrediction struct {
+	Consequence MutationType
+	CdnaPos     int // zero base
+	CdnaOffset  int
+	AaPos       int // zero base
+	AaRef       []dna.AminoAcid
+	AaAlt       []dna.AminoAcid
+}
+
+// GenomicToCdna converts genomic coordinates to cDNA coordinates. The return format is c.100+10 (HGVS).
+// The first int return is the nearest position in the coding sequence in cDNA coordinates.
+// The second int return is the distance from the nearest coding exon
+// (>0 if 5' of cds; <0 if 3' of cds, ==0 if inside coding sequence, ties break to <0).
+// Input and output positions are zero-based
+func GenomicPosToCdna(g *Gene, genomePos int) (int, int, error) {
+	var queryPos int
+	if g.posStrand { // Positive Strand
+		queryPos = genomePos - g.startPos
+	} else { // Negative Strand
+		queryPos = g.startPos - genomePos
+	}
+
+	feature := g.featureArray[queryPos]
+	switch feature {
+	case Intron:
+		var forwardOffset, reverseOffset int = 1, -1
+		for {
+			if g.featureArray[queryPos+reverseOffset] > 0 {
+				// Note: the offset values are returned with their sign flipped as we are returning
+				// the offset distance FROM the cds, not the offset distance TO the cds
+				return int(g.featureArray[queryPos+reverseOffset]), reverseOffset * -1, nil
+			}
+			if g.featureArray[queryPos+forwardOffset] > 0 {
+				return int(g.featureArray[queryPos+forwardOffset]), forwardOffset * -1, nil
+			}
+			forwardOffset++
+			reverseOffset--
+			if queryPos+forwardOffset > len(g.featureArray) || queryPos+reverseOffset < 0 {
+				return 0, 0, errors.New("no coding sequence could be found")
+			}
+		}
+
+	case UtrThree:
+		var reverseOffset int = -1
+		for g.featureArray[queryPos+reverseOffset] < 0 {
+			reverseOffset--
+			if queryPos+reverseOffset < 0 {
+				return 0, 0, errors.New("no coding sequence found before 3'UTR")
+			}
+		}
+		return int(g.featureArray[queryPos+reverseOffset]), reverseOffset * -1, nil
+
+	case UtrFive:
+		var forwardOffset int = 1
+		for g.featureArray[queryPos+forwardOffset] < 0 {
+			forwardOffset++
+			if queryPos+forwardOffset > len(g.featureArray) {
+				return 0, 0, errors.New("no coding sequence found after 5'UTR")
```

Comment:
How do noncoding RNAs fit into this?  It is ok if they don't; I am just curious.  In the GTF are they coded as all 5'UTR?  Is the Gene struct only for protein coding genes, or is there a way to have ncRNAs use a subset of the features?